SectBySectDetailPlots fails for very low participation rates

When generating `SectBySectDetailPlots` for faults with very low participation rates (zero or close to zero), we get the following Exception for a number of faults:

```
Error processing SectBySectDetailPlots plot for parent fault: Norfolk                                                                                                                                              
java.util.concurrent.ExecutionException: java.lang.IllegalStateException                                                                                                                                           
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)                                                                                                                                   
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)                                                                                                                                      
        at org.opensha.sha.earthquake.faultSysSolution.reports.plots.SectBySectDetailPlots.plot(SectBySectDetailPlots.java:198)                                                                                    
        at org.opensha.sha.earthquake.faultSysSolution.reports.ReportPageGen.generatePage(ReportPageGen.java:854)                                                                                                  
(...)
Caused by: java.lang.IllegalStateException                                                                                                                                                                         
        at com.google.common.base.Preconditions.checkState(Preconditions.java:499)                                                                                                                                 
        at org.opensha.commons.gui.plot.jfreechart.JFreeLogarithmicAxis.refreshTicksVertical(JFreeLogarithmicAxis.java:796)                                                                                        
        at org.jfree.chart.axis.LogAxis.refreshTicks(LogAxis.java:535)                                                                                                                                             
        at org.jfree.chart.axis.ValueAxis.reserveSpace(ValueAxis.java:730)                                                                                                                                         
        at org.jfree.chart.plot.XYPlot.calculateRangeAxisSpace(XYPlot.java:2785)                                                                                                                                   
        at org.jfree.chart.plot.CombinedDomainXYPlot.calculateAxisSpace(CombinedDomainXYPlot.java:388)                                                                                                             
        at org.jfree.chart.plot.CombinedDomainXYPlot.draw(CombinedDomainXYPlot.java:423)                                                                                                                           
        at org.jfree.chart.JFreeChart.draw(JFreeChart.java:1160)                                                                                                                                                   
        at org.jfree.chart.JFreeChart.createBufferedImage(JFreeChart.java:1327)                                                                                                                                    
        at org.jfree.chart.JFreeChart.createBufferedImage(JFreeChart.java:1308)                                                                                                                                    
        at org.jfree.chart.ChartUtils.writeChartAsPNG(ChartUtils.java:142)                                                                                                                                         
        at org.jfree.chart.ChartUtils.saveChartAsPNG(ChartUtils.java:268)                                                                                                                                          
        at org.jfree.chart.ChartUtils.saveChartAsPNG(ChartUtils.java:243)                                                                                                                                          
        at org.opensha.commons.gui.plot.GraphPanel.saveAsPNG(GraphPanel.java:1368)                                                                                                                                 
        at org.opensha.commons.gui.plot.GraphPanel.saveAsPNG(GraphPanel.java:1348)                                                                                                                                 
        at org.opensha.commons.gui.plot.PlotUtils.writePlots(PlotUtils.java:335)                                                                                                                                   
        at org.opensha.commons.gui.plot.PlotUtils.writePlots(PlotUtils.java:314)                                                                                                                                   
        at org.opensha.sha.earthquake.faultSysSolution.reports.plots.SectBySectDetailPlots.writeAlongStrikePlots(SectBySectDetailPlots.java:2596)                                                                  
        at org.opensha.sha.earthquake.faultSysSolution.reports.plots.SectBySectDetailPlots.getAlongStrikeLines(SectBySectDetailPlots.java:2428)                                                                    
        at org.opensha.sha.earthquake.faultSysSolution.reports.plots.SectBySectDetailPlots.buildSectionPage(SectBySectDetailPlots.java:550)
        at org.opensha.sha.earthquake.faultSysSolution.reports.plots.SectBySectDetailPlots$SectPageCallable.call(SectBySectDetailPlots.java:409)
        at org.opensha.sha.earthquake.faultSysSolution.reports.plots.SectBySectDetailPlots$SectPageCallable.call(SectBySectDetailPlots.java:385)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        ... 1 more
```

This prevents the `SectBySectDetailPlots` directory for these particular faults to be created.

The code should be able to handle these cases.

--- a/src/main/java/org/opensha/commons/gui/plot/jfreechart/JFreeLogarithmicAxis.java
+++ b/src/main/java/org/opensha/commons/gui/plot/jfreechart/JFreeLogarithmicAxis.java
@@ -962,15 +962,15 @@ public class JFreeLogarithmicAxis extends LogAxis {
 		List ticks = refreshTicks(g2, state, dataArea, edge);
 		state.setTicks(ticks);
 
-		if (ticks.size()> 20) {
+		if (ticks.size() > 15) {
 			ticks = (List) ticks.stream()
-					.filter((tick) -> ((ValueTick)tick).getTickType() == TickType.MAJOR)
+					.filter((tick) -> ((ValueTick)tick).getText().length() > 0)
 					.collect(Collectors.toList());
 		}
-		if (ticks.size()> 20) {
+		if (ticks.size() > 15) {
 			int stepSize = ticks.size() / 10;
 			ticks = IntStream.range(0, ticks.size())
-					.filter(i -> (i % stepSize) ==0)
+					.filter(i -> (i % stepSize) == 0)
 					.mapToObj(ticks::get)
 					.collect(Collectors.toList());
 		}
@@ -1257,8 +1257,8 @@ public class JFreeLogarithmicAxis extends LogAxis {
 			// if upper - lower < minRange and upper + lower < minRange, then lower might now be negative.
 			if (lower < 0) {
 				lower = SMALL_LOG_VALUE;
-				if (upper - lower < 1) {
-					upper = lower + 1;
+				if (upper - lower < minRange) {
+					upper = lower + minRange;
 				}
 			}
 

--- a/src/main/java/org/opensha/commons/gui/plot/jfreechart/JFreeLogarithmicAxis.java
+++ b/src/main/java/org/opensha/commons/gui/plot/jfreechart/JFreeLogarithmicAxis.java
@@ -103,11 +103,13 @@ import java.text.NumberFormat;
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
-import java.util.stream.Collectors;
-import java.util.stream.IntStream;
 
 import org.apache.commons.math3.util.Precision;
-import org.jfree.chart.axis.*;
+import org.jfree.chart.axis.AxisState;
+import org.jfree.chart.axis.LogAxis;
+import org.jfree.chart.axis.NumberTick;
+import org.jfree.chart.axis.Tick;
+import org.jfree.chart.axis.ValueTick;
 import org.jfree.chart.plot.Plot;
 import org.jfree.chart.plot.ValueAxisPlot;
 import org.jfree.chart.text.TextUtils;
@@ -962,19 +964,6 @@ public class JFreeLogarithmicAxis extends LogAxis {
 		List ticks = refreshTicks(g2, state, dataArea, edge);
 		state.setTicks(ticks);
 
-		if (ticks.size() > 15) {
-			ticks = (List) ticks.stream()
-					.filter((tick) -> ((ValueTick)tick).getText().length() > 0)
-					.collect(Collectors.toList());
-		}
-		if (ticks.size() > 15) {
-			int stepSize = ticks.size() / 10;
-			ticks = IntStream.range(0, ticks.size())
-					.filter(i -> (i % stepSize) == 0)
-					.mapToObj(ticks::get)
-					.collect(Collectors.toList());
-		}
-
 		Font majorTickFont = getMajorTickFont();
 		Font minorTickFont = getMinorTickFont();
 		Font powerTickFont = getPowerTickFont();
@@ -1241,6 +1230,10 @@ public class JFreeLogarithmicAxis extends LogAxis {
 			// ensure the autorange is at least <minRange> in size...
 			double minRange = getAutoRangeMinimumSize();
 			if (upper - lower < minRange) {
+				// Edge case: where <minRange> would lead us to a negative lower boundary, we adjust it.
+				if (upper + lower <= minRange / 2){
+					minRange = upper / 2;
+				}
 				upper = (upper + lower + minRange) / 2;
 				lower = (upper + lower - minRange) / 2;
 				//if autorange still below minimum then adjust by 1%
@@ -1254,15 +1247,7 @@ public class JFreeLogarithmicAxis extends LogAxis {
 				}
 			}
 
-			// if upper - lower < minRange and upper + lower < minRange, then lower might now be negative.
-			if (lower < 0) {
-				lower = SMALL_LOG_VALUE;
-				if (upper - lower < minRange) {
-					upper = lower + minRange;
-				}
-			}
-
- 			setRange(new Range(lower, upper), false, false);
+			setRange(new Range(lower, upper), false, false);
 		}
 	}
 

--- a/src/main/java/org/opensha/commons/gui/plot/jfreechart/JFreeLogarithmicAxis.java
+++ b/src/main/java/org/opensha/commons/gui/plot/jfreechart/JFreeLogarithmicAxis.java
@@ -103,13 +103,11 @@ import java.text.NumberFormat;
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
 
 import org.apache.commons.math3.util.Precision;
-import org.jfree.chart.axis.AxisState;
-import org.jfree.chart.axis.LogAxis;
-import org.jfree.chart.axis.NumberTick;
-import org.jfree.chart.axis.Tick;
-import org.jfree.chart.axis.ValueTick;
+import org.jfree.chart.axis.*;
 import org.jfree.chart.plot.Plot;
 import org.jfree.chart.plot.ValueAxisPlot;
 import org.jfree.chart.text.TextUtils;
@@ -964,6 +962,19 @@ public class JFreeLogarithmicAxis extends LogAxis {
 		List ticks = refreshTicks(g2, state, dataArea, edge);
 		state.setTicks(ticks);
 
+		if (ticks.size()> 20) {
+			ticks = (List) ticks.stream()
+					.filter((tick) -> ((ValueTick)tick).getTickType() == TickType.MAJOR)
+					.collect(Collectors.toList());
+		}
+		if (ticks.size()> 20) {
+			int stepSize = ticks.size() / 10;
+			ticks = IntStream.range(0, ticks.size())
+					.filter(i -> (i % stepSize) ==0)
+					.mapToObj(ticks::get)
+					.collect(Collectors.toList());
+		}
+
 		Font majorTickFont = getMajorTickFont();
 		Font minorTickFont = getMinorTickFont();
 		Font powerTickFont = getPowerTickFont();
@@ -1243,7 +1254,15 @@ public class JFreeLogarithmicAxis extends LogAxis {
 				}
 			}
 
-			setRange(new Range(lower, upper), false, false);
+			// if upper - lower < minRange and upper + lower < minRange, then lower might now be negative.
+			if (lower < 0) {
+				lower = SMALL_LOG_VALUE;
+				if (upper - lower < 1) {
+					upper = lower + 1;
+				}
+			}
+
+ 			setRange(new Range(lower, upper), false, false);
 		}
 	}
 

--- a/src/test/java/org/opensha/commons/gui/plot/jfreechart/JFreeChartSuite.java
+++ b/src/test/java/org/opensha/commons/gui/plot/jfreechart/JFreeChartSuite.java
@@ -1,0 +1,14 @@
+package org.opensha.commons.gui.plot.jfreechart;
+
+import org.junit.runner.RunWith;
+import org.junit.runners.Suite;
+
+@RunWith(Suite.class)
+@Suite.SuiteClasses({
+        JFreeLogarithmicAxisTest.class
+})
+public class JFreeChartSuite {
+    public static void main(String args[]) {
+        org.junit.runner.JUnitCore.runClasses(org.opensha.commons.gui.plot.jfreechart.JFreeChartSuite.class);
+    }
+}

--- a/src/test/java/org/opensha/commons/gui/plot/jfreechart/JFreeLogarithmicAxisTest.java
+++ b/src/test/java/org/opensha/commons/gui/plot/jfreechart/JFreeLogarithmicAxisTest.java
@@ -1,0 +1,58 @@
+package org.opensha.commons.gui.plot.jfreechart;
+
+import static junit.framework.Assert.*;
+
+import org.jfree.chart.axis.ValueAxis;
+import org.jfree.chart.plot.Plot;
+import org.jfree.chart.plot.PlotRenderingInfo;
+import org.jfree.chart.plot.PlotState;
+import org.jfree.chart.plot.ValueAxisPlot;
+import org.jfree.data.Range;
+import org.junit.Test;
+
+import java.awt.*;
+import java.awt.geom.Point2D;
+import java.awt.geom.Rectangle2D;
+
+public class JFreeLogarithmicAxisTest {
+
+    static class MockPlot extends Plot implements ValueAxisPlot{
+
+        final Range range;
+
+        public MockPlot(Range range) {
+            this.range = range;
+        }
+
+        @Override
+        public Range getDataRange(ValueAxis axis) {
+            return range;
+        }
+
+        @Override
+        public String getPlotType() {
+            return "";
+        }
+
+        @Override
+        public void draw(Graphics2D g2, Rectangle2D area, Point2D anchor, PlotState parentState, PlotRenderingInfo info) {
+            throw new RuntimeException("not implemented");
+        }
+    }
+
+    @Test
+    public void testSmallMinMax() {
+        JFreeLogarithmicAxis axis = new JFreeLogarithmicAxis("the axis");
+        axis.setAutoRange(false); // prevent configure() from being called when we set the Plot
+        axis.setAutoRangeMinimumSize(1, false);
+        Range range = new Range(0.00004, 0.00004);
+        axis.setPlot(new MockPlot(range));
+        axis.autoAdjustRange();
+
+        assertTrue(axis.getRange().getLowerBound() >= JFreeLogarithmicAxis.SMALL_LOG_VALUE);
+        assertTrue(axis.getRange().getUpperBound() > axis.getRange().getLowerBound());
+        assertTrue(axis.getRange().getUpperBound() - axis.getRange().getLowerBound() >= axis.getAutoRangeMinimumSize());
+
+    }
+
+}

--- a/src/test/java/org/opensha/commons/gui/plot/jfreechart/JFreeLogarithmicAxisTest.java
+++ b/src/test/java/org/opensha/commons/gui/plot/jfreechart/JFreeLogarithmicAxisTest.java
@@ -51,8 +51,11 @@ public class JFreeLogarithmicAxisTest {
 
         assertTrue(axis.getRange().getLowerBound() >= JFreeLogarithmicAxis.SMALL_LOG_VALUE);
         assertTrue(axis.getRange().getUpperBound() > axis.getRange().getLowerBound());
-        assertTrue(axis.getRange().getUpperBound() - axis.getRange().getLowerBound() >= axis.getAutoRangeMinimumSize());
 
+    }
+
+    public static void main(String[] args) {
+        new JFreeLogarithmicAxisTest().testSmallMinMax();
     }
 
 }

